### ISSUE TITLE
Add support for CUDA streams.

### ISSUE THREAD

**Issue by corbett5** [LLNL/zfp#133]
I have a project where we compute a time step on the GPU and then asynchronously copy some data back to the host for later use. This copy overlaps with the subsequent time step which saves a ton of time. Now I need to compress the data that we save, which I plan to do on device before copying it back to the CPU. It would be nice if this compression could also be asynchronous so I could overlap it with other computation.

**Comment by lindstro** [comment]
I think in principle what you're asking for would be possible via CUDA streams (I must confess to not knowing much about it), but I'm unsure how we would expose such functionality through the zfp API.  Currently the only entry point we provide is through `zfp_compress()`, which does a fair amount of setup work on the CPU and handles any data motion between CPU and GPU.  The actual CUDA compression kernel is launched some six levels deep.

Let me discuss this with our CUDA experts to see what can be done.

**Comment by lindstro** [comment]
I ran across [this paper](https://arxiv.org/pdf/2109.05410.pdf) that seems to have tackled this problem.  Not sure if their code is available.

**Comment by data-panda** [comment]
@lindstro was this something that got a place in this release (release notes does not mention so)?

**Comment by lindstro** [comment]
@data-panda No, this release does not include the latest CUDA and HIP work we have been doing.  That will end up in the next release.  Regarding CUDA streams specifically, that is not yet something our team has looked at yet.  We've had discussions with others who have looked at this (see [this paper](https://link.springer.com/chapter/10.1007/978-3-031-07312-0_1), for instance) and would welcome a contribution.

**Comment by S-o-T** [comment]
@lindstro could you please share current plans regarding CUDA support in zfp? Specifically, i am interested in:
  - user control over CUDA stream to be used for encode/decode kernel enqueuing 
  - fixed precision/accuracy and lossless compression modes

**Comment by lindstro** [comment]
We've yet to do any work on CUDA streams and lossless compression on the GPU.  It is unlikely that either would make it into the next release.  The next release will, however, have CUDA and HIP support for fixed-precision and -accuracy modes.

**Comment by S-o-T** [comment]
Thanks! Can you share an ETA for next release?

**Comment by lindstro** [comment]
I've been horrible at predicting release dates in the past and am reluctant to give false hope.  That said, we're on the hook to do a release no later than end of September.  I expect and hope it will happen well before then.

**Comment by ndcontini** [comment]
@lindstro, the MVAPICH team has implemented this both for CUDA and HIP. However, since we are not maintainers of zfp, it is likely the way we implemented it would not be release-ready. The necessary changes are relatively simple though. I tasked a junior student to do it, and he finished it relatively quickly. Below are the high-level changes needed.

1. provide a new API to pass a GPU stream. This may be as simple as adding a new member variable to one of the zfp structs and providing a function call for assigning the stream to it. It can default to the NULL stream, which is basically the current behavior of zfp. A better method (you'll see the reason why below) might be to instead have a completely different function call for asynchronous compression.
2. Remove any synchronous GPU runtime calls, e.g. `cudaMemcpy` or `cudaMemset` inside `zfp_compress` and `zfp_decompress`. There are a few calls such as these in the current code, but they can be replaced with equivalent async calls.
3. Exception to 2: one thing that cannot be replaced with an async call is the copy of the compressed size from GPU to host, which is returned by the `zfp_compress` API. In the fixed rate case, this actually does not need to be computed on the GPU, so this copy can be removed. However for non-fixed rate cases, the API needs to change such that there is a new call `zfp_compress_async` that does _not_ return the compressed size, but instead requires a call something like `zfp_synchronize` that guarantees all kernels and data movement have completed and returns the compressed size.
4. Another thing I noticed that isn't directly related but does reduce the effectiveness of zfp compression is that there is a `cudaMemset` call that initializes the output buffer of compression to all 0s. This is actually a very slow operation that is CPU driven. Ideally there would be a GPU-side (i.e. within the kernel) way to doing this initialization. If you compare profiles of compression and decompression (which does not need this memset) you can see how this effects batch compression.

Let me know if you'd like us to provide you with a proof-of-concept of how we implemented the above (sans 4).

**Comment by lindstro** [comment]
@ndcontini Thanks for all the info and suggestions.  Adding support for GPU streams has been on my to-do list, which unfortunately keeps growing every day...

We've gone to great lengths to ensure that the behavior of `zfp_compress` and `zfp_decompress` have the same semantics regardless of backend, so having `zfp_compress` return before the compressed data is available seems risky.  I'd be much more in favor of adding new async API functions, like you're suggesting for the variable-rate modes, whenever requesting asynchronous (de)compression.

The `cudaMemset` call is needed for atomics to work correctly without invoking undefined behavior when writing/updating uninitialized memory.  Of course, the CUDA specification does not say much about whether this is undefined behavior, but we're being cautious here as it clearly is UB in both C and C++.  Atomics are needed as two threads may simultaneously update the same word that straddles two consecutive zfp blocks.  For the same reason, I see no safe way of clearing memory within the compression kernel without synchronizing all threads partaking in this task before compression can begin.  I'd be surprised if that were faster than a single `cudaMemset`, which ought to be fairly optimized.  If you can think of any other approaches to this, I'm all ears.

**Comment by ndcontini** [comment]
Regarding the memset, it's possible I'm being a bit naive, but here's a bit of background info. We were implementing the changes I suggested above for HIP, and what we noticed in the profile that the kernel would not launch until the memset is done. Since for our use case we are dealing with large buffers, this results in a long memset. However there is no functional reason why some warps of would not be able to proceed with computation once the elements they are writing to are initialized. Here is my idea (but please let me know if I misunderstand something:

- thread blocks in zfp are of size 128
- In an extreme case where 64-bit elements are being compressed into 1 bit, that means the block is compressing into 8 `long long int` elements in the 1D case (seen in `writer.h`), i.e. `128 threads * 4 elements/thread * 64 bits/element / (64 / 64 bits/element) = 8`. This means that atomicity is only needed within a block, not across blocks.
- If we only need to synchronize within blocks, then we just use block level synchronization, which is finer-grain than what `cudaMemset` does, which can be thought of as kernel level synchronization. Block level synchronization would allow initialization and computation to overlap, thus reducing the end-to-end latency of a single compression.

I will admit, this is definitely "napkin arithmetic," so it's possible I need to think about this a bit more.

**Comment by lindstro** [comment]
Quick question: what is a "large buffer" to you?  I would assume single MPI messages to be small in relation to full memory usage.

Some of your arithmetic assumes certain constants in the zfp 1.0.1 implementation, some of which may change in the next release.  For instance, we're hoping to support arbitrary word sizes (8, 16, 32, and 64 bits/word).  We've also tinkered with other implementations that parallelize within a zfp block.  And you're assuming a zfp block is only 4 values (elements), which holds only in the 1D case.  In 4D, a zfp block is 4^4 = 256 values.

I'm also missing something as your expression does not come out to 8.  I think the division in the denominator should be multiplication, i.e., 64 * 64 bits/element, though I'm not sure what the first 64 denotes.

I still don't understand why atomicity would not be needed across thread blocks.  If a word straddles the substreams produced by two thread blocks, why could there not be a race condition?  In this case, both thread blocks would have to be synchronized.  Presumably one of them would be responsible for zeroing the "shared" word, but you have little control over block scheduling.

This all said, in fixed-rate mode, the zfp block size is usually a multiple of the word size, in which case none of this is needed (not even atomics).  I've been meaning to add this optimization, but it's currently far down my to-do list.

**Comment by ndcontini** [comment]
Instead of large, I suppose I could be more explicit and say that compression-based MPI protocols seem of interest to us for message sizes >4MB, thus, the memset may be operating on buffers >1MB (these are guesstimates, not so much concrete numbers).

I'm mostly considering the fixed-rate scenario, since that's what we currently support in MVAPICH. So perhaps this is a fixed-rate only optimization. I can see how for variable-rate methods you're much more likely to encounter areas where words are straddled across thread blocks, and thus what I'm proposing would not work.

**Comment by lindstro** [comment]
4 MB is still small compared to the 1+ GB datasets we often deal with.

Dealing with word-aligned blocks in fixed-rate mode ought to be straightforward, and I could see us implementing that for the next release.  In _d_ = 1 dimension, this currently requires you to choose a rate among {16, 32, 48, 64} since the word size is fixed at 64 bits.

When _d_ = 1 and _rate_ = 16, one could also further optimize the code as each compressed block is exactly one word.  This could be done without even accessing memory until the single 64-bit word (accumulated in a register) is output, with no need for atomics or inner-loop tests to check if a word is ready for output. Such an optimization would require more extensive code changes, but ought to improve performance quite a bit for this special but important case.

In _d_ = 2 dimensions, rates would be restricted to {4, 8, 12, ..., 64}, and in _d_ = 3 dimensions to {1, 2, 3, ..., 64}.  With support for smaller word sizes, granularity would be finer in all dimensions.
